tools/codegen/dest/register_dispatch_key.py

```diff
@@ -257,6 +261,9 @@ def gen_class_set_output_body(self, k: SchemaKind) -> str:
                 elif self.dispatch_key == DispatchKey.CUDA:
                     empty_impl = "at::native::empty_cuda"
                     empty_strided_impl = "at::native::empty_strided_cuda"
+                elif self.dispatch_key == DispatchKey.DefaultBackend:
+                    empty_impl = "at::empty"
```

Comment:
> Totally agreed that this makes the most sense to unblock native functions with DefaultBackend functional/inplace variants for structured kernels.

This is a misunderstanding of what the patch does. Although there is a test for dispatch key for DefaultBackend here, this code is exercised even if a structured kernel doesn't have any DefaultBackend in the dispatch dictionary in native functions yaml. An easy way to see this is add, which is exercised by the tests in this PR, doesn't have DefaultBackend in the dispatch dictionary:

```
- func: add.out(Tensor self, Tensor other, *, Scalar alpha=1, Tensor(a!) out) -> Tensor(a!)
  structured: True
  structured_inherits: TensorIteratorBase
  dispatch:
    CPU, CUDA: add_out
    SparseCPU: add_out_sparse_cpu
    SparseCUDA: add_out_sparse_cuda
    MkldnnCPU: mkldnn_add_out
```

> So right now, a custom backend would need to implement their own version of add/add_/add_out, to get full add functionality, but only sin_out to get everything for sin.

This is factually incorrect. Both add and sin (after @hameerabbasi's PR) are structured, and in fact in both cases you only need to implement `add_out` and `sin_out` to get full functionality. If you registered a specialized implementation you can get it a little faster (from skipped redispatches), but it is not necessary.

It is true that it is a bit arbitrary whether or not something is structured or not to say what you need to implement. The fix for that is to port more stuff to structured.

Let me know if there is something else you were getting at here that I missed. I think there is generally a question about "what should we expect backends to implement" and in some cases (e.g., XLA) I don't think it is an improvement to only have to implement the out version; they'd rather implement the functional version. But we can only pick one direction to do the defaulting the way the dispatcher is currently setup, need some more fancy gear otherwise.

Reply:
Thanks for writing that out! Staring at the codegen output from this PR also helped a little. Just to lay it out, what I had in my head was that we had to choose between two options (taking `sin` as an example):
- Register a specialized version of `sin` that skips redispatches, but requires separate implementations for other backends
- Register a backend-agnostic version of `sin`. Backends don't need to implement their own version, but it requires extra dispatches

The structured kernel codegen now just picks **both** choices, and uses the fact that `DefaultBackend` is an alias key to ensure that "the right thing" happens. If another backend registers their own specialized implementation, the dispatcher will give that implementation precedence. Otherwise, we default to the `DefaultBackend` implementation.



You probably already had this in mind with [this issue](https://github.com/pytorch/pytorch/issues/50953), but we can probably do something similar with overloads- potentially also for overloads that don't necessarily call each other for performance reasons. For example, I started a port of `pow`, and it has an overload `pow.Tensor_Scalar(Tensor self, Scalar exponent) -> Tensor` that's implemented directly as a unary op, even though it could technically be written a little less efficiently to call out to the more general Tensor overload. I could imagine doing something similar to here, where we keep the specialized implementation of the overload for perf reasons, but also codegen a slower, backend-agnostic variant of it that calls the more general Tensor overload.

Reply:
@ezyang @bdhirsh just to check my understanding, I think this PR narrows the BC-breakage window substantially but doesn't entirely close it. (Not that it should, just want to make sure I'm thinking about all the moving parts properly.)

E.g. if we have 
1. an unstructured op with dispatch configured a la `sin` in `native_functions.yaml`:
```
- func: foo ...
  dispatch:
    DefaultBackend: foo

- func: foo_ ...
  dispatch:
    DefaultBackend: foo_

- func: foo.out ...
  dispatch:
    CPU, CUDA: foo_out
```

2. where `foo` and `foo_` kernels really do work for all backends

3. and an out-of-tree backend provider relies on this support for `foo` and `foo_`

4. and the provider doesn't care about, and has not registered a kernel for, `foo.out`

Then before this PR, moving `foo`/`foo_`/`foo.out` to SK would suddenly oblige the backend provider to register kernels for `foo` and `foo_` to keep the same level of support it had previously gotten for free. 

After this PR, the generated DefaultBackend support means this is no longer necessary (though still allowed), but IIUC the backend provider will still need to have registered a kernel for `foo.out` for the op to actually run, correct?

Reply:
@bhosmer I don't know how to square (4) with (3).

> and the provider doesn't care about, and has not registered a kernel for, foo.out

OK, so the provider hasn't registered foo.out. So.... how exactly was the DefaultBackend `foo` and `foo_` implemented? Can't be by calling `foo.out`, because if they actually did that then it would be broken if they hadn't registered `foo.out`. So you're saying that `foo` in some sort of other way that doesn't use out... to which I say, "Don't do that, and structured kernels will make you not do that."

Reply:
Right, I'm just confirming that this PR assumes that dependency between unstructured variants, and that it's possible for an unstructured op to break the assumption while still coloring inside the lines, registration-wise. I don't think we should try to handle this case - I have zero ideas for real-world ops that would actually use it - but wanted to confirm my understanding of exactly which configurations this PR guarantees BC for.